forc-pkg/src/source/ipfs.rs

```diff
@@ -0,0 +1,242 @@
+use crate::{
+    manifest::{self, PackageManifestFile},
+    source,
+};
+use anyhow::Result;
+use futures::TryStreamExt;
+use ipfs_api::IpfsApi;
+use ipfs_api_backend_hyper as ipfs_api;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{
+    fmt,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+use tar::Archive;
+use tracing::{info, warn};
+
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct Cid(cid::Cid);
+
+/// A client that can interact with local ipfs daemon.
+pub type IpfsClient = ipfs_api::IpfsClient;
+
+/// Package source at a specific content address.
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
+pub struct Source(pub Cid);
+
+/// A pinned instance of an ipfs source
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize, Deserialize)]
+pub struct Pinned(pub Cid);
+
+impl Pinned {
+    pub const PREFIX: &'static str = "ipfs";
+}
+
+const PUBLIC_GATEWAY: &str = "https://ipfs.io";
+const IPFS_DIR_NAME: &str = "ipfs";
+const IPFS_CACHE_DIR_NAME: &str = "cache";
+
+impl FromStr for Cid {
+    type Err = <cid::Cid as FromStr>::Err;
+
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
+        let cid = s.parse()?;
+        Ok(Self(cid))
+    }
+}
+
+impl source::Pin for Source {
+    type Pinned = Pinned;
+    fn pin(&self, _ctx: source::PinCtx) -> Result<(Self::Pinned, PathBuf)> {
+        let cid = &self.0;
+        let pinned = Pinned(cid.clone());
+        let path = pkg_cache_dir(cid);
+        Ok((pinned, path))
+    }
+}
+
+impl source::Fetch for Pinned {
+    fn fetch(&self, ctx: source::PinCtx, repo_path: &Path) -> Result<PackageManifestFile> {
+        // TODO: implement local cache search for ipfs sources.
```

Comment:
nit: Is there an issue for this? It would be good to put the issue number in the TODO comment